src/services/completions.ts

```diff
@@ -658,6 +670,225 @@ function isCheckedFile(sourceFile: SourceFile, compilerOptions: CompilerOptions)
+                        break;
+                    case "string":
+                        elements.push(factory.createStringLiteral(type.value));
+                        break;
+                }
+            }
+        }
+        if (elements.length === 0) {
+            return undefined;
+        }
+
+        const newClauses = map(elements, element => {
+            return factory.createCaseClause(element, []);
+        });
+        const printer = createSnippetPrinter({
+            removeComments: true,
+            module: options.module,
+            target: options.target,
+            newLine: getNewLineKind(getNewLineCharacter(options, maybeBind(host, host.getNewLine))),
+        });
+        const insertText = formatContext
+            ? printer.printAndFormatSnippetList(
+                ListFormat.MultiLine | ListFormat.NoTrailingNewLine,
+                factory.createNodeArray(newClauses),
+                sourceFile,
+                formatContext)
+            : printer.printSnippetList(
+                ListFormat.MultiLine | ListFormat.NoTrailingNewLine,
+                factory.createNodeArray(newClauses),
+                sourceFile);
+
+        const firstClause = printer.printSnippetList(ListFormat.SingleLine, factory.createNodeArray([first(newClauses)!]), sourceFile);
+        return {
+            entry: {
+                name: `${firstClause} ...`,
+                kind: ScriptElementKind.unknown,
+                sortText: SortText.GlobalsOrKeywords,
+                insertText,
+                hasAction: importAdder.hasFixes() || undefined,
+                source: CompletionSource.SwitchCases,
+            },
+            importAdder,
+        };
+    }
+
+    return undefined;
+}
+
+interface CaseClauseTracker {
+    addValue(value: string | number): void;
+    hasValue(value: string | number | PseudoBigInt): boolean;
+}
+
+function newCaseClauseTracker(checker: TypeChecker, clauses: readonly (CaseClause | DefaultClause)[]): CaseClauseTracker {
+    const existingStrings = new Set<string>();
+    const existingNumbers = new Set<number>();
+    const existingBigInts = new Set<string>();
+
+    for (const clause of clauses) {
+        if (!isDefaultClause(clause)) {
+            if (isLiteralExpression(clause.expression)) {
+                const expression = clause.expression;
+                switch (expression.kind) {
+                    case SyntaxKind.NoSubstitutionTemplateLiteral:
+                    case SyntaxKind.StringLiteral:
+                        existingStrings.add(expression.text);
+                        break;
+                    case SyntaxKind.NumericLiteral:
+                        existingNumbers.add(parseInt(expression.text));
+                        break;
+                    case SyntaxKind.BigIntLiteral:
+                        const parsedBigInt = parseBigInt(endsWith(expression.text, "n") ? expression.text.slice(0, -1) : expression.text);

```

Comment:
Remind me why we need to parse and then re-stringify bigints?

Reply:
We need to use a standard representation of big ints for our sets, so that big ints that are written differently but are the same end up with the same representation. Since we can't use actual big ints because they're not supported in all runtimes (that might change eventually?), we use "pseudo big ints" in our codebase to represent big ints, but then that's not something we can put into a set, therefore I need to stringify the pseudo big ints.